File: keyfilter.py
#!/usr/bin/env python

import sys

REPLACE_CONTENT = [
    ('USERNAME', 'github'),
    ('PASSWORD', '123456')
]

def usage():
    print("""usage:
        1. for git smudge
            python keyfilter.py smudge
        2. for git clean
            python keyfilter.py clean
        """)

def smudge():
    for line in sys.stdin:
        for a in REPLACE_CONTENT:
            line = line.replace(*a[::-1])
        sys.stdout.write(line)

def clean():
    for line in sys.stdin:
        for a in REPLACE_CONTENT:
            line = line.replace(*a)
        sys.stdout.write(line)

if __name__ == '__main__':
    try:
        if sys.argv[1] == 'smudge':
            smudge()
        elif sys.argv[1] == 'clean':
            clean()
        else:
            usage()
    except Exception as e:
        print(e)
        usage()

File: keyfilter.js
#!/usr/bin/env node

const REPLACE_CONTENT = {
    'USERNAME': 'github.com',
    'PASSWORD': '123456'
}

process.stdin.setEncoding('utf8');

function usage() {
    console.log(`usage:
        1. for git smudge
            keyfilter.js smudge
        2. for git clean
            keyfilter.js clean
    `);
}

function smudge() {
    process.stdin.on('readable', () => {
        let chunk = process.stdin.read();
        if (chunk !== null) {
            for(let key in REPLACE_CONTENT) {
                chunk = chunk.replace(REPLACE_CONTENT[key], key);
            }
            process.stdout.write(`${chunk}`);
        }
    });
}

function clean() {
    process.stdin.on('readable', () => {
        let chunk = process.stdin.read();
        if (chunk !== null) {
            for(let key in REPLACE_CONTENT) {
                chunk = chunk.replace(key, REPLACE_CONTENT[key]);
            }
            process.stdout.write(`${chunk}`);
        }
    });
}

function main() {
    try {
        if (process.argv.length >= 3) {
            switch(process.argv[2]) {
                case 'smudge':
                    smudge();
                    break;
                case 'clean':
                    clean();
                    break;
                default:
                    usage();
            }
        } else {
            usage();
        }
    } catch (error) {
        console.log(error);
        usage();
    }
}

main()
